grafonnet-base/veneer/variable.libsonnet

```diff
@@ -0,0 +1,463 @@
+        var.withName(name)
+        + var.withType('query')
+        + var.withQuery(query),
+
+      '#withDatasource':: d.func.new(
+        'Select a datasource for the variable template query.',
+        args=[
+          d.arg('type', d.T.string),
+          d.arg('uid', d.T.string),
+        ]
+      ),
+      withDatasource(type, uid):
+        var.datasource.withType(type)
+        + var.datasource.withUid(uid),
+
+      '#withDatasourceFromVariable':: d.func.new(
+        'Select the datasource from another template variable.',
+        args=[
+          d.arg('variable', d.T.object),
+        ]
+      ),
+      withDatasourceFromVariable(variable):
+        if variable.type == 'datasource'
+        then self.withDatasource(variable.query, '${%s}' % variable.name)
+        else error "`variable` not of type 'datasource'",
+
+      '#withRegex':: d.func.new(
+        |||
+          `withRegex` can extract part of a series name or metric node segment. Named
+          capture groups can be used to separate the display text and value
+          ([see examples](https://grafana.com/docs/grafana/latest/variables/filter-variables-with-regex#filter-and-modify-using-named-text-and-value-capture-groups)).
+        |||,
+        args=[
+          d.arg('value', d.T.string),
+        ]
+      ),
+      withRegex(value): {
+        regex: value,
+      },
+
+      '#withSort':: d.func.new(
+        |||
+          Choose how to sort the values in the dropdown.
+
+          This can be called as `withSort(<number>) to use the integer values for each
+          option. If `i==0` then it will be ignored and the other arguments will take
+          precedence.
+
+          The numerical values are:
+
+          - 1 - Alphabetical (asc)
+          - 2 - Alphabetical (desc)
+          - 3 - Numerical (asc)
+          - 4 - Numerical (desc)
+          - 5 - Alphabetical (case-insensitive, asc)
+          - 6 - Alphabetical (case-insensitive, desc)
+        |||,
+        args=[
+          d.arg('i', d.T.number, default=0),
+          d.arg('type', d.T.string, default='alphabetical'),
+          d.arg('asc', d.T.boolean, default=true),
+          d.arg('caseInsensitive', d.T.boolean, default=false),
+        ],
+      ),
+      withSort(i=0, type='alphabetical', asc=true, caseInsensitive=false):
+        if i != 0  // provide fallback to numerical value
+        then { sort: i }
+        else
+          {
+            local mapping = {
+              alphabethical:
+                if !caseInsensitive
+                then
+                  if asc
+                  then 1
+                  else 2
+                else
+                  if asc
+                  then 5
+                  else 6,
+              numerical:
+                if asc
+                then 3
+                else 4,
+            },
+            sort: mapping[type],
+          },
+
+      // TODO: Expand with Query types to match GUI
+      queryTypes: {
```

Comment:
Query Types are currently a front-end only concept, they get translated to their equivalent query. To make them available here we'd need to do a bit more reverse engineering.